libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs

```diff
@@ -110,70 +134,148 @@ public static bool IsTrustedServiceUrl(string serviceUrl)
+                                // update the cache with completed task so all new requests will get it
+                                tokenTaskCache[TokenCacheKey] = task;
+                            }
+                            else
+                            {
+                                // it failed, shorten the refresh time for another task to try again in a 30s
+                                autoRefreshTimes[TokenCacheKey] = DateTime.UtcNow + TimeSpan.FromSeconds(30);
+                            }
+                        });
+                }
+            }
+        }
+
+        /// <summary>
+        /// Do not call this except from GetTokenRefreshAsync() 
+        /// </summary>
+        /// <returns></returns>
+        private Task<OAuthResponse> _getCurrentTokenTask(bool forceRefresh)
+        {
+            Task<OAuthResponse> oAuthTokenTask;
+
+            // if there is not a task or we are forcing it
+            if (forceRefresh || tokenTaskCache.TryGetValue(TokenCacheKey, out oAuthTokenTask) == false)
             {
-                _cache[TokenCacheKey] = token;
+                // create it
+                oAuthTokenTask = RefreshTokenAsync();
+                tokenTaskCache[TokenCacheKey] = oAuthTokenTask;
+
+                // set initial refresh time
+                autoRefreshTimes[TokenCacheKey] = DateTime.UtcNow + AutoTokenRefreshTimeSpan;
+            }
+            // if task is in faulted or canceled state then replace it with another attempt
+            else if (oAuthTokenTask.IsFaulted || oAuthTokenTask.IsCanceled)
+            {
+                oAuthTokenTask = RefreshTokenAsync();
+                tokenTaskCache[TokenCacheKey] = oAuthTokenTask;
+
+                // set initial refresh time
+                autoRefreshTimes[TokenCacheKey] = DateTime.UtcNow + AutoTokenRefreshTimeSpan;
             }
 
-            return token.access_token;
+            return oAuthTokenTask;
         }
 
         private bool ShouldSetToken(HttpRequestMessage request)
         {
-            if (IsTrustedUrl(request.RequestUri))
+            if (TrustedUri(request.RequestUri))
             {
                 return true;
             }
 
+            System.Diagnostics.Debug.WriteLine($"Service url {request.RequestUri.Authority} is not trusted and JwtToken cannot be sent to it.");
             return false;
         }
 
-        private static bool IsTrustedUrl(Uri uri)
+        private static bool TrustedUri(Uri uri)
         {
-            lock (_trustedHostNamesSync)
+            DateTime trustedServiceUrlExpiration;
+            if (TrustedHostNames.TryGetValue(uri.Host, out trustedServiceUrlExpiration))
             {
-                if (_trustedHostNames.TryGetValue(uri.Host, out DateTime trustedServiceUrlExpiration))
+                // check if the trusted service url is still valid
+                if (trustedServiceUrlExpiration > DateTime.UtcNow.Subtract(TimeSpan.FromMinutes(5)))
```

Comment:
I don't understand the 5 minute timespan here. It's... unexpected. Is this a workaround for clockskew issues? It will lead to code that says "trusted" even when a service url has passed the current system time. That seems a mistake.

Reply:
No idea.

---
In reply to: [198678137](https://github.com/Microsoft/botbuilder-dotnet/pull/702#discussion_r198678137) [](ancestors = 198678137)